### PR TITLE
Add homogenization constraints to enforce the Hill-Mandel conditions + some volume average conditions

### PR DESCRIPTION
**Pull request by reverendbedford**
Adds a system for enforcing Hill-Mandel conditions + some volume average constraints.  This requires you to setup periodic constraints using the standard MOOSE system and then use the new scalar kernel and associated user object to enforce the volume average conditions.  PR includes an action to do this all automatically for you.

### CODE REVIEW

File: tests/homogenization/small-tests/1d-strain.i [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+[]
+
+[Mesh]
+  type = FileMesh
+  file = '1d.exo'
+[]
+
+[MeshModifiers]
+  [./left]
+    type = SideSetsFromPoints
```
Comment (arovinelli) [inline tests/homogenization/small-tests/1d-strain.i:18]: change to SideSetsFromPointsGenerator

File: tests/homogenization/small-tests/1d-strain.i [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+    new_boundary = 'left'
+    points = '-1 0 0'
+  [../]
+
+  [./right]
+    type = SideSetsFromPoints
```
Comment (arovinelli) [inline tests/homogenization/small-tests/1d-strain.i:24]: change to SideSetsFromPointsGenerator

File: tests/homogenization/small-tests/1d-stress.i [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+[]
+
+[Mesh]
+  type = FileMesh
+  file = '1d.exo'
+[]
+
+[MeshModifiers]
+  [./left]
+    type = SideSetsFromPoints
```
Comment (arovinelli) [inline tests/homogenization/small-tests/1d-stress.i:18]: change to SideSetsFromPointsGenerator

File: tests/homogenization/small-tests/1d-stress.i [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+    new_boundary = 'left'
+    points = '-1 0 0'
+  [../]
+
+  [./right]
+    type = SideSetsFromPoints
```
Comment (arovinelli) [inline tests/homogenization/small-tests/1d-stress.i:24]: change to SideSetsFromPointsGenerator

File: src/actions/NEMLMechanicsAction.C [hunk 5]
```diff
@@ -48,9 +52,19 @@ InputParameters NEMLMechanicsAction::validParams() {
       "Names of the eigenstrains");
 
   params.addParam<std::vector<SubdomainName>>(
-      "block", "The list of subdomain names where neml mehcanisc must be used, "
-               "default all blocks. Helpful when different physiscs and "
-               "kernels are required on different blocks.");
+      "block", "The list of subdomain names where neml mehcanisc should be used, "
```
Comment (arovinelli) [inline src/actions/NEMLMechanicsAction.C:55]: could you correct spelling here

File: src/kernels/TotalStressDivergenceNEML.C [hunk 6]
```diff
@@ -151,3 +184,153 @@ TotalStressDivergenceNEML::largeDeformationGeoJac(
...
+           - _inv_def_grad[_qp](k,a) * _inv_def_grad[_qp](b,j));
+
+      for (unsigned int m = 0; m < _ndisp; m++) {
+        for (unsigned int n = 0; n < _ndisp; n++) {
+          value +=
+              _detJ[_qp] * _material_jacobian[_qp](i,j,m,n) * 
+              _df[_qp](m,a) * _inv_def_grad[_qp](b,n) * 
+              _grad_test[_i][_qp](k) * _inv_def_grad[_qp](k,j);
+        }
+      }
```
Comment (arovinelli) [inline src/kernels/TotalStressDivergenceNEML.C:250]: remove

File: src/kernels/TotalStressDivergenceNEML.C [hunk 7]
```diff
@@ -151,3 +184,153 @@ TotalStressDivergenceNEML::largeDeformationGeoJac(
...
+    }
+  }
```
Comment (arovinelli) [inline src/kernels/TotalStressDivergenceNEML.C:252]: remove {}, out of all you {} only for k loop

File: src/materials/ComputeNEMLStressBase.C [hunk 8]
```diff
@@ -117,6 +120,14 @@ void ComputeNEMLStressBase::computeQpProperties() {
...
   _energy[_qp] = u_np1;
   _dissipation[_qp] = p_np1;
+
+  // Do some postprocessing
+  if (_ld) {
+    _PK[_qp] = _J[_qp] * _stress[_qp] * _F_inv[_qp].transpose();
+  }
+  else {
+    _PK[_qp] = _stress[_qp];
+  }
```
Comment (arovinelli) [inline src/materials/ComputeNEMLStressBase.C:130]: remove extra {}

File: src/scalarkernels/HomogenizationConstraintScalarKernel.C [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+    _ncomps(HomogenizationConstants::required.at(_ld)[_ndisp-1]),
+    _integrator(getUserObject<HomogenizationConstraintIntegral>("integrator")),
+    _indices(HomogenizationConstants::indices.at(_ld)[_ndisp-1]),
+    _residual(_integrator.getResidual()),
+    _jacobian(_integrator.getJacobian())
+{
+  if (_var.order() != _ncomps) {
+    mooseError("Homogenization kernel requires a variable of order ",
+               _ncomps);
+  }
```
Comment (arovinelli) [inline src/scalarkernels/HomogenizationConstraintScalarKernel.C:47]: remove {}

File: src/userobjects/HomogenizationConstraintIntegral.C [hunk 10]
```diff
@@ -0,0 +1,251 @@
...
+      getParam<std::vector<FunctionName>>("targets");
+  
+  unsigned int nfns = names.size();
+  if (nfns != _ncomps) {
+    mooseError("Homogenization constraint user object needs ",
+               _ncomps, " functions");
+  }
+  for (unsigned int i = 0; i < nfns; i++) {
+    const Function * const f = &getFunctionByName(names[i]);
+    if (!f) mooseError("Function ", names[i], " not found.");
```
Comment (arovinelli) [inline src/userobjects/HomogenizationConstraintIntegral.C:58]: I'm not suer if you need this

File: src/userobjects/HomogenizationConstraintIntegral.C [hunk 11]
```diff
@@ -0,0 +1,251 @@
...
+      for (_hh = 0; _hh < _ncomps; _hh++) {
+        unsigned int a = _indices[_hh].first;
+        unsigned int b = _indices[_hh].second;
+        if (_ctypes[_h] == HomogenizationConstants::ConstraintType::Stress) {
+          if (_ld) {
+            res(i,j,a,b) = ldStressJacobian(i,j,a,b);
+          }
+          else {
+            res(i,j,a,b) = sdStressJacobian(i,j,a,b);
+          }
```
Comment (arovinelli) [inline src/userobjects/HomogenizationConstraintIntegral.C:186]: remove {}

File: src/userobjects/HomogenizationConstraintIntegral.C [hunk 12]
```diff
@@ -0,0 +1,251 @@
...
+        }
+        else if (_ctypes[_h] == HomogenizationConstants::ConstraintType::Strain) {
+          if (_ld) {
+            res(i,j,a,b) = ldStrainJacobian(i,j,a,b);
+          }
+          else {
+            res(i,j,a,b) = sdStrainJacobian(i,j,a,b);
+          }
```
Comment (arovinelli) [inline src/userobjects/HomogenizationConstraintIntegral.C:194]: remove {}

File: src/userobjects/HomogenizationConstraintIntegral.C [hunk 13]
```diff
@@ -0,0 +1,251 @@
...
+        }
+        else {
+          mooseError("Unknown constraint type in Jacobian calculator!");
+        }
```
Comment (arovinelli) [inline src/userobjects/HomogenizationConstraintIntegral.C:198]: remove  {}

File: src/userobjects/HomogenizationConstraintIntegral.C [hunk 14]
```diff
@@ -0,0 +1,251 @@
...
+      else if (_ctypes[_h] == HomogenizationConstants::ConstraintType::Strain) {
+        Real f = (_indices[_h].first == _indices[_h].second) ? 1.0 : 0.0;
+        res(_indices[_h].first,_indices[_h].second) = 
+            0.5*(_F[_qp](_indices[_h].first,_indices[_h].second) +
+                 _F[_qp](_indices[_h].second,_indices[_h].first)) -
+            (f + _targets[_h]->value(_t, _q_point[_qp]));
+      }
+      else {
+        mooseError("Unknown constraint type in the integral!");
+      }
```
Comment (arovinelli) [inline src/userobjects/HomogenizationConstraintIntegral.C:162]: remove {}

File: src/userobjects/HomogenizationConstraintIntegral.C [hunk 15]
```diff
@@ -0,0 +1,251 @@
...
+    }
+    else {
+      if (_ctypes[_h] == HomogenizationConstants::ConstraintType::Stress) {
+        res(_indices[_h].first,_indices[_h].second) = 
+            _stress[_qp](_indices[_h].first,_indices[_h].second) - 
+            _targets[_h]->value(_t, _q_point[_qp]);
+      }
```
Comment (arovinelli) [inline src/userobjects/HomogenizationConstraintIntegral.C:152]: remove {}

File: src/userobjects/HomogenizationConstraintIntegral.C [hunk 16]
```diff
@@ -0,0 +1,251 @@
...
+      else if (_ctypes[_h] == HomogenizationConstants::ConstraintType::Strain) {
+        Real f = (_indices[_h].first == _indices[_h].second) ? 1.0 : 0.0;
+        res(_indices[_h].first,_indices[_h].second) = 
+            _F[_qp](_indices[_h].first,_indices[_h].second) - 
+            (f + _targets[_h]->value(_t, _q_point[_qp]));
+      }
+      else {
+        mooseError("Unknown constraint type in the integral!");
+      }
```
Comment (arovinelli) [inline src/userobjects/HomogenizationConstraintIntegral.C:145]: remove {}

File: src/userobjects/HomogenizationConstraintIntegral.C [hunk 17]
```diff
@@ -0,0 +1,251 @@
...
+  RankTwoTensor res;
+  res.zero();
+
+  for (_h = 0; _h < _ncomps; _h++) {
+    if (_ld) {
+      if (_ctypes[_h] == HomogenizationConstants::ConstraintType::Stress) {
+        res(_indices[_h].first,_indices[_h].second) = 
+            _PK1[_qp](_indices[_h].first,_indices[_h].second) - 
+            _targets[_h]->value(_t, _q_point[_qp]);
+      }
```
Comment (arovinelli) [inline src/userobjects/HomogenizationConstraintIntegral.C:136]: remove {}

File: src/userobjects/HomogenizationConstraintIntegral.C [hunk 18]
```diff
@@ -0,0 +1,251 @@
...
+  Real res = 0.0;
+  for (unsigned int k = 0; k < _ndisp; k++) {
+    res += _J[_qp] * _stress[_qp](i,k) * (_invF[_qp](b,a) * _invF[_qp](j,k) - 
+                                          _invF[_qp](j,a) * _invF[_qp](b,k));
+    for (unsigned int s = 0; s < _ndisp; s++) {
+      for (unsigned int t = 0; t < _ndisp; t++) {
+        res += _J[_qp] * _material_jacobian[_qp](i,k,s,t) * _df[_qp](s,a) *
+            _invF[_qp](b,t) * _invF[_qp](j,k);
+      }
+    }
```
Comment (arovinelli) [inline src/userobjects/HomogenizationConstraintIntegral.C:237]: remove {}

File: src/kernels/TotalStressDivergenceNEML.C [hunk 19]
```diff
@@ -26,8 +33,26 @@ TotalStressDivergenceNEML::TotalStressDivergenceNEML(const InputParameters &para
...
+    // Check the order of the scalar variable
+    unsigned int needed = HomogenizationConstants::required.at(_ld)[_ndisp-1];
+    if (_macro_gradient->order() != needed)
+      mooseError("The homogenization macro gradient variable must have order ",
+                 needed);
+    
+    // Check the number of constraints
+    auto types = getParam<std::vector<std::string>>("constraint_types");
+    if (types.size() != needed)
+      mooseError("The kernel must be supplied ", needed, " constraint types.");
```
[the remark this answers is not in the document]
Comment (reverendbedford) [reply src/kernels/TotalStressDivergenceNEML.C:53]: Yes?  _macro_gradient.order() == needed == types.size().

File: tests/homogenization/small-tests/1d-strain.i (hunk 1 above)
Comment (reverendbedford) [reply tests/homogenization/small-tests/1d-strain.i:18]: Did this everywhere.

**Pull request merged**